Phenicx Anechoic / Multitrack

nkundiushuti opened the pull request:
I have created a new version of the #345 

PHENICX-Anechoic dataset - multi-source dataset for orchestral music with note annotations (used in score-informed source separation, note transcription)
https://zenodo.org/record/840025
 - [x] Changed core.py, validate.py to work with multitrack 
- [x] Create a script in scripts/, e.g. make_my_dataset_index.py
- [x]  Run the script on the canonical version of the dataset and save the index in mirdata/indexes/ e.g. my_dataset_index.json
- [X] Create a module in mirdata, e.g. mirdata/my_dataset.py
 - [X] Create tests in tests/, e.g. test_my_dataset.py
 - [X] Run tests
 - [X] Add the module to docs/source/mirdata.rst
 - [X] Add the module to docs/source/table.rst

rabitt commented on mirdata/core.py:
```diff
@@ -203,6 +205,19 @@ def _track(self, track_id):
+        Args:
+            mtrack_id (str): mtrack id of the multitrack
+        Returns:
+            multitrack (dataset.MultiTrack): an instance of this dataset's MultiTrack object
```
nit: the syntax sphinx expects for Returns is (strangely)

```
Returns:
    type: description of type
```

rabitt commented on mirdata/core.py:
```diff
@@ -391,6 +419,48 @@ class MultiTrack(Track):
+        self._data_home = data_home
+        self._multitrack_paths = index["multitracks"][mtrack_id]
+
+        if metadata and mtrack_id in metadata:
```
I just changed this slightly (most recent commit on master) so that it lazy loads the metadata, instead of loading it when the object is created.
suggested change:
```
@property
def _multitrack_metadata(self):
    if metadata and mtrack_id in metadata:
        return metadata[mtrack_id]
    elif metadata:
        return metadata
    return None
```

rabitt commented on mirdata/core.py:
```diff
@@ -216,6 +231,19 @@ def load_tracks(self):
+        """
+        return {mtrack_id: self.multitrack(mtrack_id) for mtrack_id in self.mtrack_ids}
+
     def choice_track(self):
```
do we want to add a choice_multitrack? or is it overkill?

nkundiushuti replied:
it could be done.

rabitt reviewed (commented):
Left some nits, but this structure looks great

nkundiushuti commented:
@rabitt I have added a multitrack attribute to the Dataset class, similarly to the track class
multitrack_class (mirdata.core.Multitrack or None): a Multitrack class 
I think this was needed for docs.

rabitt commented on mirdata/core.py:
```diff
@@ -391,6 +446,51 @@ class MultiTrack(Track):
+
+        self._data_home = data_home
+        self._multitrack_paths = index["multitracks"][mtrack_id]
+        self._multitrack_metadata = multitrack_metadata
```
Sorry I missed some small details in my last comments - 

Line 483 should be `self._metadata = metadata`

Then, before line 488, add:
`metadata = self._metadata()`

and all the `mtrack_id` variables --> `self.mtrack_id`

rabitt commented on mirdata/core.py:
```diff
@@ -391,6 +446,51 @@ class MultiTrack(Track):
 
     """
 
+    def __init__(
```
For the end of the init, should we also add:

```
self._index = index
self.track_ids = self._index["multitracks"]["tracks"]
```
and maybe two additional properties, one for getting all associated track objects:
```
@property
def tracks(self):
    return {t: Track(t, self.data_home, self.dataset_name, self._index, self._metadata) for t in self.track_ids}
```

and one defining the audio property to use when mixing, which in core we can default to None, but we can overwrite in the loaders
```
@property
def track_audio_property(self):
    raise NotImplementedError("Mixing is not possible in this dataset")
```

With that we can remove the `_check_mixable` function everywhere in the code below.

nkundiushuti commented on mirdata/validate.py:
```diff
@@ -76,12 +76,12 @@ def validate_files(file_dict, data_home, verbose):
     missing = {}
     invalid = {}
     for file_id, file in tqdm.tqdm(file_dict.items(), disable=not verbose):
-        # multitrack case
```
@rabitt, this gave me an error in validating the index we agreed upon so I had to move it one level down. 'mozart' being a key in the multitrack which has the subkeys 'tracks' (and maybe 'audio-mix').

rabitt replied:
Nice. maybe this function deserves a refactor in the future. It's getting a little nest-y

nkundiushuti commented on mirdata/datasets/phenicx_anechoic.py:
```diff
@@ -0,0 +1,413 @@
+        """
+        return self.get_target(self.sections[section])
+
+    def get_notes_target(self, track_keys, notes_property='notes'):
```
a similar function to get_target, which mixes notes for various tracks

nkundiushuti commented on mirdata/datasets/phenicx_anechoic.py:
```diff
@@ -0,0 +1,413 @@
+        self.piece = self.mtrack_id
+
+    @property
+    def tracks(self):
```
I had to reimplement this otherwise the Track object created would have been the one from core.py

rabitt replied:
^^ see my fix in the comment in `core.py`. I think with that fix you can delete the reimplementing

nkundiushuti commented on mirdata/datasets/phenicx_anechoic.py:
```diff
@@ -0,0 +1,413 @@
+        return {t: Track(t, self._data_home, self._dataset_name, self._index, self._metadata) for t in self.track_ids}
+
+    @property
+    def track_audio_property(self):
```
without doing any other tricks, this was the only way to set this attribute for my class

rabitt replied:
yes! this is how I intended for it to be set. Maybe it needs to be more clear e.g. in contributing?

nkundiushuti commented:
@rabitt,  I am about to implement the tests. could you take a look to see if everything is ok?

rabitt commented on mirdata/core.py:
```diff
@@ -386,17 +452,69 @@ class MultiTrack(Track):
+
+    @property
+    def tracks(self):
+        return {t: Track(t, self._data_home, self._dataset_name, self._index, self._metadata) for t in self.track_ids}
```
you're right - there's a bug here. I think we need to pass `_track_class` to the init of the `MultiTrack` class like we do in the `Dataset` class, and then here do 
`{t: self._track_class(t, self._data_home, self._dataset_name, self._index, self._metadata) for t in self.track_ids}`

rabitt commented on mirdata/datasets/phenicx_anechoic.py:
```diff
@@ -0,0 +1,413 @@
+        #### the attribute of Track which returns the relevant audio file for mixing
+        return 'audio'
+
+    def get_audio_instrument(self,instrument):
```
nit: maybe call this `get_audio_for_instrument`

rabitt commented on mirdata/datasets/phenicx_anechoic.py:
```diff
@@ -0,0 +1,413 @@
+        """
+        return getattr(self.tracks[self.instruments[instrument]], self.track_audio_property)[0]
+
+    def get_audio_section(self,section):
```
similar nit: `get_audio_for_section`

rabitt commented on mirdata/datasets/phenicx_anechoic.py:
```diff
@@ -0,0 +1,413 @@
+
+        return annotations.NoteData(intervals, values)
+
+    def get_notes_instrument(self,instrument, notes_property='notes'):
```
same nit for these two `get_notes_for_*`

rabitt reviewed (commented):
Left some comments!

nkundiushuti commented:
> Left some comments!

thanks! I implemented the changes. I will go on with the tests

nkundiushuti commented:
this currently fails on mypy and it's not very clear for me why
do you have any clue, @rabitt @magdalenafuentes ?
this is where it fails:
```
    @property
    def audio(self) -> Optional[Tuple[np.ndarray, float]]:
        """the track's audio

        Returns:
            * np.ndarray - the mono audio signal
            * float - The sample rate of the audio file

        """
        assert self.audio_paths is not None
        audio_mix, sr = load_audio(self.audio_paths[0])

        for i in range(1, self.n_voices):
            audio, _ = load_audio(self.audio_paths[i])
            audio_mix += audio
        return audio_mix, sr
```

rabitt commented on mirdata/datasets/phenicx_anechoic.py:
```diff
@@ -0,0 +1,421 @@
+            * float - The sample rate of the audio file
+
+        """
+        assert self.audio_paths is not None and isinstance(self.audio_paths,list)
```
from the init, audio_paths will never be None, but it could be an empty list. @drubinstein suggested that this assert could be what's causing mypy to flag it. Try removing the assert!

drubinstein replied:
A bit of a hack, but you could replace the load audio lines with
```
        audio_mix, sr = cast(Tuple[np.ndarray, float], load_audio(self.audio_paths[0]))

        for i in range(1, self.n_voices):
            audio, _ = cast(Tuple[np.ndarray, float], load_audio(self.audio_paths[i]))
      ```
      
      Alternatively (and maybe more correct), you could use a temporary variable. Something like 
      ```
      x = load_audio(self.audio_paths[i]
      audio, _ = x
      ```
      
      Going to point out that this code will break if one audio path cannot be opened or is None as you can't add None to an np.ndarray (nor can you add None to None)

nkundiushuti replied:
thanks @drubinstein ! the first solution worked while the second hasn't. so I have done that

nkundiushuti commented:
it doesn't work with the assert either. I put the assert to make sure the
self.audio_paths is not None because the error is: None object is not
iterable.

În joi, 11 feb. 2021 la 17:43, Rachel Bittner <notifications@github.com> a
scris:

> *@rabitt* commented on this pull request.
> ------------------------------
>
> In mirdata/datasets/phenicx_anechoic.py
> <https://github.com/mir-dataset-loaders/mirdata/pull/461#discussion_r574654903>
> :
>
> > +        self.notes_path = os.path.join(self._data_home, self._track_paths["notes"][0])
> +
> +        self.notes_original_path = os.path.join(
> +            self._data_home, self._track_paths["notes_original"][0]
> +        )
> +
> +    @property
> +    def audio(self) -> Optional[Tuple[np.ndarray, float]]:
> +        """the track's audio
> +
> +        Returns:
> +            * np.ndarray - the mono audio signal
> +            * float - The sample rate of the audio file
> +
> +        """
> +        assert self.audio_paths is not None and isinstance(self.audio_paths,list)
>
> from the init, audio_paths will never be None, but it could be an empty
> list. @drubinstein <https://github.com/drubinstein> suggested that this
> assert could be what's causing mypy to flag it. Try removing the assert!
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mir-dataset-loaders/mirdata/pull/461#pullrequestreview-588763219>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJG62YPAQSKEBTYQ3WDUM3S6QCLPANCNFSM4XAXL6WQ>
> .
>


-- 
Marius Miron
http://mariusmiron.com/

nkundiushuti commented:
@rabitt, all yours for review!

rabitt commented on mirdata/datasets/phenicx_anechoic.py:
```diff
@@ -0,0 +1,422 @@
+
+        self.n_voices = len(self.audio_paths)
+
+        self.notes_path = os.path.join(self._data_home, self._track_paths["notes"][0])
```
remember to update these if we merge #468 first!

rabitt replied:
@nkundiushuti Looks like we should indeed update these. I'm happy to commit to your branch with the fix if you're sick of this PR :)

rabitt commented on mirdata/datasets/phenicx_anechoic.py:
```diff
@@ -0,0 +1,422 @@
+
+    def get_audio_voice(self, id_voice: int) -> Optional[Tuple[np.ndarray, float]]:
+        """the track's audio
+
```
```
Args:
    id_voice (int): ...
```

rabitt replied:
Dont forget this one

rabitt commented on mirdata/datasets/phenicx_anechoic.py:
```diff
@@ -0,0 +1,422 @@
+        data_home,
+        dataset_name,
+        index,
+        track_class=Track,
```
I wouldn't make these defaults, I'd directly pass them: `__init__(self, ...., Track, None)`

rabitt commented on mirdata/datasets/phenicx_anechoic.py:
```diff
@@ -0,0 +1,422 @@
+            np.ndarray: instrument audio with shape (n_samples, n_channels)
+
+        """
+        assert instrument in self.instruments.keys()
```
nit: I prefer raising errors over asserts, so the user gets a more clear message if something goes wrong, and its easier to test for.

rabitt commented on tests/test_core.py:
```diff
@@ -63,20 +117,35 @@ def test_dataset_errors():
     with pytest.raises(NotImplementedError):
         d.load_tracks()
 
+    with pytest.raises(KeyError):
```
hmm weird that this would raise a `KeyError`, while `load_tracks` raises a `NotImplementedError`. Seems like a (small) bug. I'd expect them to have the same behavior

rabitt commented on tests/test_core.py:
```diff
@@ -63,20 +117,35 @@ def test_dataset_errors():
     with pytest.raises(NotImplementedError):
         d.choice_track()
 
+    with pytest.raises(KeyError):
```
same here - i'd expect the same behavior as `choice_track`

rabitt commented on tests/test_core.py:
```diff
@@ -87,41 +156,66 @@ class TestMultiTrack1(core.MultiTrack):
             self._data_home = data_home
+            self.track_ids = []
+            self.mtrack_ids = []
+            self._metadata = None
```
`self._metadata = lambda: None`

rabitt commented on tests/test_core.py:
```diff
@@ -87,41 +156,66 @@ class TestMultiTrack1(core.MultiTrack):
         mtrack.get_target(["a"])
 
-    with pytest.raises(NotImplementedError):
+    with pytest.raises(AssertionError):
```
if we replace the asserts above that I commented on, these can change

rabitt commented on tests/test_core.py:
```diff
@@ -87,41 +156,66 @@ class TestMultiTrack1(core.MultiTrack):
         mtrack.get_mix()
 
+    with pytest.raises(AttributeError):
+        mtrack._multitrack_medata
```
`medata` --> `metadata`?

this should also raise a `NotImplementedError`

rabitt commented on tests/test_core.py:
```diff
@@ -132,8 +226,16 @@ class TestMultiTrack(core.MultiTrack):
         def __init__(self, mtrack_id, data_home):
             self.mtrack_id = mtrack_id
             self._data_home = data_home
-            self.tracks = {t: TestTrack(t) for t in ["a", "b", "c"]}
-            self.track_audio_property = "f"
+            self._dataset_name = "foo"
+            self._index = None
+            self._metadata = None
+            self._track_class = TestTrack
+            self.track_ids = ["a", "b", "c"]
```
now that you simplified the multitrack class, i think you can remove this init from the tests, and just pass in a simple index like `{'multitracks': {'mtrack_id': {'tracks': ['a', b', 'c']}}}`

I wrote similar tests for Track in #468

rabitt commented on tests/test_phenicx_anechoic.py:
```diff
@@ -0,0 +1,304 @@
+        "audio": tuple,
+    }
+
+    run_track_tests(track, expected_attributes, expected_property_types)
```
should we write a similar function for `run_multitrack_tests`? We might even be able to use the same, just replacing `track` by `multitrack` in the function call.

rabitt replied:
nevermind! I just saw you already did it :)

nkundiushuti commented on mirdata/core.py:
```diff
@@ -203,6 +213,29 @@ def _track(self, track_id):
+                self.data_home,
+                self.name,
+                self._index,
+                lambda: self._metadata,
```
good catch. I basically replicated the _track object and I think there are problems there too
```
            return self._track_class(
                track_id,
                self.data_home,
                self.name,
                self._index,
                lambda: self._metadata,
            )
```
should have also dataset_name which is not a property of class Dataset(object):

rabitt replied:
i think that does pass `dataset_name` - it's `Dataset.name`! 

```
            return self._track_class(
                track_id,
                self.data_home,
                self.name,  # <-- this is dataset_name
                self._index,
                lambda: self._metadata,
            )
```

nkundiushuti replied:
yes! you're right :) I will update this accordingly

genisplaja commented:
Really sorry for the merging delay @nkundiushuti of the #468, it is now done!

nkundiushuti commented:
> Really sorry for the merging delay @nkundiushuti of the #468, it is now done!

no problem! thank you!

nkundiushuti commented:
@rabitt, I have addressed your comments in the previous comments. please let me know if I need to change something else.

rabitt commented:
> @rabitt, I have addressed your comments in the previous comments. please let me know if I need to change something else.

Looks like there were a few little comments that got missed from the previous review. Let me know if you want me to do some of it, I'm happy to help!

nkundiushuti commented:
> > @rabitt, I have addressed your comments in the previous comments. please let me know if I need to change something else.
> 
> Looks like there were a few little comments that got missed from the previous review. Let me know if you want me to do some of it, I'm happy to help!

uh I don't know what I missed? some error raising in test_core.py? could you please take a look?

rabitt commented:
Ooops I just noticed the docs aren't building properly - going to push a fix and then lets merge!

rabitt commented:
@nkundiushuti feel free to merge! And sorry this was such a marathon

nkundiushuti commented:
> @nkundiushuti feel free to merge! And sorry this was such a marathon

thanks to you for the patience and for the help!
